src/main.cpp

```diff
@@ -7638,6 +7641,7 @@ bool SendMessages(const Consensus::Params& params, CNode* pto)
             // Only actively request headers from a single peer, unless we're close to today.
             if ((nSyncStarted == 0 && fFetch) || pindexBestHeader->GetBlockTime() > GetTime() - 24 * 60 * 60) {
                 state.fSyncStarted = true;
+                state.nHeadersSyncTimeout = GetTimeMicros() + HEADERS_DOWNLOAD_TIMEOUT_BASE + HEADERS_DOWNLOAD_TIMEOUT_PER_HEADER * (EstimateNetHeight(params, pindexBestHeader->nHeight, pindexBestHeader->GetBlockTime()) - pindexBestHeader->nHeight);
```

Comment:
This looks like a correct adaptation to Zcash (with block target spacing that changes at Heartwood activation). :+1:

Reply:
I think it was Blossom upgrade that introduced new block target spacing. Maybe HEADERS_DOWNLOAD_TIMEOUT_BASE of 15 minutes and HEADERS_DOWNLOAD_TIMEOUT_PER_HEADER of 1 millisecond per header still need some tweaking, considering the fact that Zcash block header is larger (contains equihash solution).

Reply:
Yes, Bitcoin headers are 80 bytes and Zcash headers are 1488 bytes. I'd suggest doubling `HEADERS_DOWNLOAD_TIMEOUT_BASE` to 30 minutes, and setting `HEADERS_DOWNLOAD_TIMEOUT_BASE` to 10 ms (it probably doesn't need to scale quite linearly, which would give 18.6 ms).

Reply:
Hmm, this would be a ~13 day timeout at the current chain height. Is that too long?

Reply:
![Screenshot from 2022-11-19 11-49-06](https://user-images.githubusercontent.com/37712022/202855993-364f5957-a080-4a79-be09-41ac24341894.png)
I did a couple of tests and the average downloading time per header did not exceed 1.5ms.
Since the purpose of timeout is preventing node from waiting forever in case of stalling peer, I guess it's better to under-estimate these constants than to over-estimate them. Imho, if we under-estimate the timeout, the worst that can happen is early switching to the other peer as a source of the remaining headers, which is not necessarily so bad as waiting.

Reply:
Also, I'm leaning towards reducing `HEADERS_DOWNLOAD_TIMEOUT_BASE` and maybe compensating that reduction by raising `HEADERS_DOWNLOAD_TIMEOUT_PER_HEADER` slightly. Rationale for this would be seemingly better scaling regarding the sync starting height.

Reply:
I disagree that it's better to underestimate, because if they're too low then there is a risk that *no* peers will be able to provide the headers, or even worse, that only an adversary's peers will.

I suggest using 10ms as the timeout; I'll restore my suggestion that did that. We can always reduce it in a future release.

Reply:
@str4d and I concluded that the worst case was likely dominated by the round-trip latency of `getheaders`. The peer we're downloading from could be anywhere in the world, and the number of headers downloaded in a single `getheaders` message is `MAX_HEADERS_RESULTS = 160`, so we need to allow (RTT latency)/160 for `HEADERS_DOWNLOAD_TIMEOUT_PER_HEADER`. 10ms should be easily enough.